File: genui/02_ag_ui/step_02_ag_ui_tools_and_state/json_patch.py
"""JSON Patch (RFC 6902), the three operations this step's state deltas
use: add, replace and remove, with JSON Pointer paths such as
/dashboard/metrics/- (append) or /dashboard/table. An unknown op, a path
into nothing or a list index that is not a number raises ValueError.

The server applies each patch to its own copy of the state before sending
it, so server and page hold the same object after every STATE_DELTA.
json-patch.mjs is the same code for the page.
"""


def split_pointer(path):
    """'/a/b~1c/0' -> ['a', 'b/c', '0']. An empty path means the root."""
    if path == "":
        return []
    return [part.replace("~1", "/").replace("~0", "~") for part in path[1:].split("/")]


OPS = ("add", "replace", "remove")


def index(part):
    """A list index: digits only, so 'x' is refused instead of being coerced."""
    if not part.isdigit():
        raise ValueError(f"not a list index: {part!r}")
    return int(part)


def walk(document, parts):
    """The container the last part points into, and that last part."""
    target = document
    for part in parts[:-1]:
        try:
            target = target[index(part)] if isinstance(target, list) else target[part]
        except (KeyError, IndexError, TypeError):
            raise ValueError(f"no such path: {part!r}") from None
    return target, parts[-1]


def apply_patch(document, operations):
    """Apply the operations in order, in place. Returns the document."""
    for operation in operations:
        op, parts = operation.get("op"), split_pointer(operation.get("path", ""))
        if op not in OPS:
            raise ValueError(f"unsupported op {op}")
        if not parts:
            raise ValueError("the root cannot be patched in place")
        parent, key = walk(document, parts)
        if isinstance(parent, list):
            if op == "add":
                if key == "-":
                    parent.append(operation["value"])
                else:
                    parent.insert(index(key), operation["value"])
            elif op == "replace":
                parent[index(key)] = operation["value"]
            else:
                del parent[index(key)]
        elif isinstance(parent, dict):
            if op == "remove":
                parent.pop(key, None)
            else:
                parent[key] = operation["value"]
        else:
            raise ValueError(f"no such path: {operation.get('path')!r}")
    return document

File: genui/02_ag_ui/step_02_ag_ui_tools_and_state/json-patch.mjs
// JSON Patch (RFC 6902) for the page: add, replace and remove, with JSON
// Pointer paths. The same code as json_patch.py, so server and page agree.
// The patch is data from the wire, so the path is checked before it is
// followed: no prototype keys, list indexes are digits, a missing parent
// is an error rather than a crash further down.

const OPS = new Set(["add", "replace", "remove"]);
const FORBIDDEN = new Set(["__proto__", "constructor", "prototype"]); // would reach Object.prototype

export function splitPointer(path) {
  if (path === "") return [];
  return path.slice(1).split("/").map((part) => part.replace(/~1/g, "/").replace(/~0/g, "~"));
}

function index(part) {
  if (!/^\d+$/.test(part)) throw new Error(`not a list index: ${part}`);
  return Number(part);
}

function step(parent, part) {
  if (FORBIDDEN.has(part)) throw new Error(`refused path segment ${part}`);
  const child = Array.isArray(parent) ? parent[index(part)] : Object.hasOwn(parent, part) ? parent[part] : undefined;
  if (child === undefined || child === null || typeof child !== "object") throw new Error(`no such path: ${part}`);
  return child;
}

export function applyPatch(document, operations) {
  for (const operation of operations) {
    if (!OPS.has(operation.op)) throw new Error(`unsupported op ${operation.op}`);
    const parts = splitPointer(operation.path ?? "");
    if (parts.length === 0) throw new Error("the root cannot be patched in place");
    let parent = document;
    for (const part of parts.slice(0, -1)) parent = step(parent, part);
    const key = parts[parts.length - 1];
    if (FORBIDDEN.has(key)) throw new Error(`refused path segment ${key}`);
    if (Array.isArray(parent)) {
      if (operation.op === "add") key === "-" ? parent.push(operation.value) : parent.splice(index(key), 0, operation.value);
      else if (operation.op === "replace") parent[index(key)] = operation.value;
      else parent.splice(index(key), 1);
    } else if (operation.op === "remove") delete parent[key];
    else parent[key] = operation.value;
  }
  return document;
}
